### ISSUE TITLE
Doesn't work for merge commits

### ISSUE THREAD

**Issue by ffMathy** [8BitJonny/gh-get-current-pr#16]
If the latest commit is a merge commit, it can't find the pull request associated.

**Comment by 8BitJonny** [comment]
Can you show me your .yml file? It sounds to me like you have passed the flag `filterOutClosed: true` to the action which tells the action to ignore any closed Pull Requests.

I went through this Repo's Actions and here I found a Merge Commit on which it successfully detected the associated PR
https://github.com/8BitJonny/gh-get-current-pr/runs/2536503872
So it works in my repo with this workflow config:
```yml
on: [push]

jobs:
  test_job:
    runs-on: ubuntu-latest
    name: Returns the PR number
    steps:
      - name: Checkout
        uses: actions/checkout@v2
      - name: Get PR
        uses: ./
        id: getPR
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
      - name: Display PR properties
        run: |
          echo "The PR Number is ${{ steps.getPR.outputs.number }}"
          echo "The PR Title is ${{ steps.getPR.outputs.pr_title }}"
          echo "The PR Url is ${{ steps.getPR.outputs.pr_url }}"
          echo "The PR Created_At is ${{ steps.getPR.outputs.pr_created_at }}"
          echo "The PR Merged_At is ${{ steps.getPR.outputs.pr_merged_at }}"
          echo "The PR Closed_At is ${{ steps.getPR.outputs.pr_closed_at }}"
```

**Comment by ffMathy** [comment]
I am not using `filterOutClosed: true`. I can't share the YML file, because it is covered by an NDA.

However, here are the areas of importance:

```
      - uses: actions/checkout@v2.3.4

      - uses: 8BitJonny/gh-get-current-pr@master
        name: Get current pull request (if present)
        id: PR
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}

      - name: Generate constants
        id: constants
        shell: bash
        run: |
          label="${{fromJson(steps.PR.outputs.pr || '{}').labels[0].name}}"
          echo "::set-output name=label::${label}"

      - name: Read constants
        shell: bash
        run: |
          echo ${{ steps.constants.outputs.label }}
```

For merge commits, it returns empty string :shrug:

**Comment by 8BitJonny** [comment]
That snippet is already helpful 👍🏼 
Have you tried adding a debugging step, to see if the PR really wasn't found or if your `Generate constants` doesn't find a label? Something like that?
```yml
- name: Debug GetPR output
   run: |
      echo ${{ steps.PR.outputs.number }}
```

Anyway, I will also try to replicate your issue in the coming days on my machine to see, what the issue might be

**Comment by ffMathy** [comment]
Yeah I have such a step. It still yields nothing.

**Comment by ffMathy** [comment]
@8BitJonny had time to look into this?

**Comment by 8BitJonny** [comment]
I'm currently in my last 10 days of writing my bachelor thesis. So sadly no.
If you find anything else about this issue pls share. I will look into it once everything is done on the university side. Sorry to keep you waiting

**Comment by ffMathy** [comment]
@8BitJonny how did your bachelor's thesis go? I hope it went as you had expected.

Hereby just a friendly reminder that this is still an issue. Looking forward to your help :pray:

And no rush!

**Comment by ffMathy** [comment]
It seems that `context.sha` is incorrect.

If I change the steps to this:

```
      - uses: actions/checkout@v2.3.4
      
      - name: Get commit SHA
        id: sha
        shell: bash
        run: |
          echo ::set-output name=commit::$(git rev-list --no-merges -n 1 HEAD)

      - uses: 8BitJonny/gh-get-current-pr@master
        name: Get current pull request (if present)
        id: PR
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          sha: ${{ steps.sha.outputs.commit }}

      - name: Generate constants
        id: constants
        shell: bash
        run: |
          label="${{fromJson(steps.PR.outputs.pr || '{}').labels[0].name}}"
          echo "::set-output name=label::${label}"

      - name: Read constants
        shell: bash
        run: |
          echo ${{ steps.constants.outputs.label }}
```

In other words, by changing the SHA to a SHA without merge commits, it works always. Even for brand new PRs with the label set.

**Comment by 8BitJonny** [comment]
@ffMathy Sorry for the late reply. Just passed my final Bachelor Exam two days ago, so sorry for the delay.

Back to the issue at hand:
I'm glad you found a workaround! I can see why that works because the Merge Commit (for which the PR lookup does not work) will then be skipped. But as I already said, this feels like a temporary workaround to me. 
Or is this topic now done for you?

If not, can you tell me the exact steps you take when merging two branches?
Imagine you have a production branch called `main` and a branch called `feature-1`. Do you merge via Github.com? And if so which merge strategy do you use (Squash and Merge, Rebase and Merge, or Merge Commit)? Or if you do it over the command line, what exact commands do you do?

The most helpful scenario would be if you could give some kind of mini repo in which you can demonstrate the bug, because I am still not able to reproduce the same bug that you have.

Still glad, you found a workaround. Have a sunny day

**Comment by ffMathy** [comment]
Hi @8BitJonny and thank you so much for getting back to me! Glad you passed! Congratulations!

Well the whole issue actually happens just by creating a pull request that has a label already from the beginning.

For some reason, it then points to a merge commit.

If I then make a dummy push afterwards, your solution works.

It's very strange indeed. It has nothing to do with actually merging anything. It only happens when building pull requests, and only when the pull request is initially created :shrug:

So your action doesn't work at all on initial pull request creation. Only if a dummy push is made afterwards.

I think that may be an issue for others. But I can't see how you can fix it via the API, unless you do a similar "hack" that I did.

**Comment by 8BitJonny** [comment]
@ffMathy Are you able to create a dummy repo for that to show it to me in action?
That would be super helpful

**Comment by 8BitJonny** [comment]
Or maybe can you please just have a look at #17 and the action run https://github.com/8BitJonny/gh-get-current-pr/runs/3625564479?check_suite_focus=true

You mean that case? Because I just now see that after creating the PR over the Github UI, the first commit doesn't yield an PR Number.

**Comment by 8BitJonny** [comment]
The second commit in #17 yields a PR number again as can be seen in https://github.com/8BitJonny/gh-get-current-pr/runs/3625595738?check_suite_focus=true

Interesting 🤔

**Comment by 8BitJonny** [comment]
Okay wait a second.

You say the issue is that the first commit in a PR is not recognized as a commit belonging to a PR, correct?

Because I just realized that this is logical since:

1. You first create a commit
2. Push the commit to github (if commit was created locally and not through Github.com)
3. The Action is automatically triggered for that new commit before you can do anything else (like create a PR for that commit)
4. Then you create a PR for that new Commit, so that the commit is listed as the first commit in that new PR. But the commit and its action was run BEFORE the PR was created, hence the action is not able to display any PR number at the time of execution

Does that make sense?

**Comment by ffMathy** [comment]
Almost. Except the PR also triggers an action build. I actually also have a PR trigger.

And that fails. The latest commit after creating a PR is a merge commit for some reason. Not sure why github does this 😅

You can try it here in this repo. Make a GitHub actions file that has both a push trigger and a pull request trigger.

Then make a commit to a new branch. You'll see github actions trigger a build. This won't have a PR (for obvious reasons).

But then you make a PR for the branch. The GitHub actions triggers again. However, now, for some reason, the latest commit will be a merge commit, and it can't find the PR. This is the problem, and the weird part.

But then, if you make a commit again, triggering a new build, the pull request is again recognized.

**Comment by 8BitJonny** [comment]
And that's the absolute most important crux of it all.
That piece of information would have been nice to have 18 messages ago 😄

**Comment by ffMathy** [comment]
Ah wow yeah I thought I had shared that at the beginning :joy:

So sorry :joy.

**Comment by 8BitJonny** [comment]
😄 
Anyway, with that piece of information I found this [article](https://frontside.com/blog/2020-05-26-github-actions-pull_request/#how-does-pull_request-affect-actionscheckout) that explains a bit why the sha env is different when the `pull_request` trigger is used. 

You can even see a detailed `git log` in [one of my test PRs](https://github.com/8BitJonny/gh-get-current-pr/pull/21/checks?check_run_id=3630521380)

This is something that I cannot control. But the article already gives a very solid workaround like this:
```
- uses: 8BitJonny/gh-get-current-pr@master
        name: Get current pull request (if present)
        id: PR
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          sha: ${{ github.event.pull_request.head.sha }}
```

and the last line will specify the commit sha correctly no matter if `push` or `pull_request` triggered the action run.

I will add that bit to the ReadMe and will then consider this issue as resolved 👍🏼
